lib/cn/csched_sp3_work.c

```diff
@@ -163,34 +166,49 @@ sp3_work_wtype_root(
     runlen_max = thresh->rspill_runlen_max;
     runlen = 1;
 
+    znode = cn_kvset_can_zspill(le->le_kvset, rmap);
+
     /* Look for a contiguous sequence of non-busy kvsets.
-     *
-     * TODO: Starting with the first non-busy kvset, count the number of
-     * kvsets contiguous from the first that would all spill to the same
-     * leaf node.
      */
     while ((le = list_prev_entry_or_null(le, le_link, &tn->tn_kvset_list))) {
         if (kvset_get_workid(le->le_kvset) != 0)
             break;
 
+        if (znode) {
+            struct cn_tree_node *n = cn_kvset_can_zspill(le->le_kvset, rmap);
+
+            if (n) {
+                if (n->tn_nodeid == znode->tn_nodeid && runlen < runlen_max) {
+                    ++runlen;
+                    continue;
+                } else {
+                    assert(runlen && runlen <= runlen_max);
+                    *action = CN_ACTION_ZSPILL;
+                    *rule = CN_RULE_ZSPILL;
+                    return runlen;
+                }
+            }
+        }
+
+        /* TODO: Stop if we see a zspill candidate.
```

Comment:
Is there a ticket for this

Reply:
No. I was afraid it may cause issues in a pathological case where every other kvset is a zspill candidate. Maybe I'll get rid of the todo and we can work on this if it's an issue.